internal/rotate.go

```diff
@@ -113,6 +113,34 @@ func (h *CertRotationHandler) Rotate() error {
 	return os.Rename(h.stateFile, h.stateFile+".completed")
 }
 
+// ResumeRotation checks if we have an incomplete cert rotation. If so, it
+// will attempt to complete this rotation. The main reason this would happen
+// is if a power failure occurred during `.Rotate`
+func (h *CertRotationHandler) ResumeRotation(online bool) error {
+	if !online {
+		// There's not much we can do because most rotation steps require
+		// network access. However `fioconfig extract` runs at early boot and
+		// needs help with one specific condition: the finalizeStep was able to
+		// update sota.toml but didn't update config.encrypted. In this case
+		// we can complete that one step locally and be good.
+		if h.State.DeviceConfigUpdated && !h.State.Finalized {
+			log.Print("Incomplete certificate rotation state found. Will attempt to complete")
+			step := finalizeStep{}
+			if err := step.finalizeConfigEncrypted(h); err != nil {
```

Comment:
IIUC there is no warranty that the new `sota.toml` was saved in this case.

In fact, there are two sub-use-cases:
1. Device config was updated, new sota.toml was saved, new config.encrypted was not saved - this is what this code handles well.
2. Device config was updated, new sota.toml and new config.encrypted were not saved - this is where this code will fail to work correctly. The reason is that sota.toml still references to old keys, while config.encrypted will be now encrypted with new keys.

IMHO in order to mitigate this weird race condition state we need to split the `finalize` step into two parts: a sotaTomlUpdated (which saves the new keys and sota.toml) plus an actual finalizeStep (which only saves the new config.encrypted).

Reply:
> The finalize step does two things that we should try and treat as "atomic"

This is exactly what does not hold true:
1. The `h.State.DeviceConfigUpdated` is set when device config was uploaded. If it is `true` - we don't know if `sota.toml` was saved or not.
2. The `h.State.Finalized` is set when the `config.encrypted` was saved. If it is `false` - we don't know if `sota.toml` was saved or not.
3. The `finalizeConfigEncrypted` only saves new `config.encrypted`, it does not save new `sota.toml`.

That's why I say that we need a separate boolean value to signal that `sota.toml` was already saved, but `config.encrypted` wasn't yet saved. A power outage can happen in-between these two operations.